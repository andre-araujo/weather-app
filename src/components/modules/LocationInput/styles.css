.container {
  width: 100vw;
  height: 100vh;
  overflow: auto;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  padding: 50px;
  background: rgba(255, 255, 255, 0.9);
  position: fixed;
  z-index: 10;
  top: 0;
  left: 0;
}

.input {
  font-size: 40px;
  outline: none;
  border: none;
  border-bottom: 5px solid #333;
  background: none;
  transition: .2s border-color;
  text-align: center;
  width: 100%;
  max-width: 400px;
}

.input:focus {
  border-color: orange;
}

.icon {
  margin-right: 10px;
}

.row {
  margin-bottom: 15px;
}
