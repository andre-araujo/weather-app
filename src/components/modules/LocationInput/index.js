import React, { Component } from 'react';
import { func, string } from 'prop-types';

import styles from './styles.less';
import Icon from '../../elements/Icon';
import Button from '../../elements/Button';
import getLatLong from '../../../lib/getLatLong';

class LocationInput extends Component {
  state = {
    location: '',
  }

  handleInput = (e) => {
    this.setState({
      location: e.target.value,
    });
  }

  handleGetCoords = () => {
    const { onGetCoords } = this.props;

    getLatLong(({ coords }) => {
      onGetCoords({
        latitude: coords.latitude,
        longitude: coords.longitude,
      });
    }, (error) => {
      onGetCoords({
        error,
      });
    });
  }

  handleSearch = (e) => {
    const { onSearch } = this.props;
    e.preventDefault();

    onSearch(this.state.location);
  }

  render() {
    const { location } = this.state;

    return (
      <form className={styles.container} onSubmit={this.handleSearch}>
        <div className={styles.row}>
          <input className={styles.input} type="text" onChange={this.handleInput} value={location} placeholder="Ex: Rio de Janeiro" />
        </div>
        {this.props.error && (
          <div className={styles.row}>
            <p className={styles.error}>{this.props.error}</p>
          </div>
        )}
        <div className={styles.row}>
          <Button type="button" onClick={this.handleGetCoords} data-test="LocationInput_GetCoords">
            <span className={styles.icon}><Icon.Compass /></span>
            get location
          </Button>
        </div>
        <div className={styles.row}>
          <Button type="submit">Search</Button>
        </div>
      </form>
    );
  }
}

LocationInput.defaultProps = {
  error: null,
};

LocationInput.propTypes = {
  error: string,
  onSearch: func.isRequired,
  onGetCoords: func.isRequired,
};

export default LocationInput;
