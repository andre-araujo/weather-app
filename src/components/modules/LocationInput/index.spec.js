import React from 'react';
import { shallow } from 'enzyme';

import LocationInput from './index';

jest.mock('../../../lib/getLatLong');

const getLatLongMock = require('../../../lib/getLatLong').default;

let props;

beforeEach(() => {
  props = {
    error: null,
    onSearch: jest.fn(),
    onGetCoords: jest.fn(),
  };
});

describe('LocationInput component', () => {
  it('should match snapshot with no error message', () => {
    expect(shallow(<LocationInput {...props} />)).toMatchSnapshot();
  });

  it('should match snapshot with error message', () => {
    props.error = 'my error message';
    expect(shallow(<LocationInput {...props} />)).toMatchSnapshot();
  });

  it('should preventDefault when form is submited', () => {
    const preventDefault = jest.fn();

    const wrapper = shallow(<LocationInput {...props} />);
    wrapper
      .find('form')
      .simulate('submit', { preventDefault });

    expect(preventDefault).toHaveBeenCalledTimes(1);
  });

  it('should call onSearch when form is submited', () => {
    const preventDefault = jest.fn();

    const wrapper = shallow(<LocationInput {...props} />);
    wrapper
      .find('form')
      .simulate('submit', { preventDefault });

    expect(props.onSearch).toHaveBeenCalledTimes(1);
  });

  it('should call onSearch with input value when form is submited', () => {
    const preventDefault = jest.fn();

    const wrapper = shallow(<LocationInput {...props} />);

    wrapper
      .find('input')
      .simulate('change', { target: { value: 'my input value' } });

    wrapper
      .find('form')
      .simulate('submit', { preventDefault });

    expect(props.onSearch).toHaveBeenCalledWith('my input value');
  });

  it('should call onGetCoords with getLatLong result when LocationInput_GetCoords is clicked', () => {
    const wrapper = shallow(<LocationInput {...props} />);

    getLatLongMock.mockImplementation(callback => callback({
      coords: {
        latitude: 10,
        longitude: 10,
      },
    }));

    wrapper
      .find('[data-test="LocationInput_GetCoords"]')
      .simulate('click');

    expect(props.onGetCoords).toHaveBeenCalledWith({
      latitude: 10,
      longitude: 10,
    });
  });

  it('should call onGetCoords with getLatLong error when LocationInput_GetCoords is clicked', () => {
    const wrapper = shallow(<LocationInput {...props} />);

    getLatLongMock.mockImplementation((_, callback) => callback(new Error('error')));

    wrapper
      .find('[data-test="LocationInput_GetCoords"]')
      .simulate('click');

    expect(props.onGetCoords).toHaveBeenCalledWith({ error: Error('error') });
  });
});
