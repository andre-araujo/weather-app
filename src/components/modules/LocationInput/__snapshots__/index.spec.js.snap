// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationInput component should match snapshot with error message 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      onChange={[Function]}
      placeholder="Ex: Rio de Janeiro"
      type="text"
      value=""
    />
  </div>
  <div>
    <p>
      my error message
    </p>
  </div>
  <div>
    <Button
      data-test="LocationInput_GetCoords"
      onClick={[Function]}
      type="button"
    >
      <span>
        <Component />
      </span>
      get location
    </Button>
  </div>
  <div>
    <Button
      data-test={null}
      onClick={null}
      type="submit"
    >
      Search
    </Button>
  </div>
</form>
`;

exports[`LocationInput component should match snapshot with no error message 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      onChange={[Function]}
      placeholder="Ex: Rio de Janeiro"
      type="text"
      value=""
    />
  </div>
  <div>
    <Button
      data-test="LocationInput_GetCoords"
      onClick={[Function]}
      type="button"
    >
      <span>
        <Component />
      </span>
      get location
    </Button>
  </div>
  <div>
    <Button
      data-test={null}
      onClick={null}
      type="submit"
    >
      Search
    </Button>
  </div>
</form>
`;
